src/app-layer-imap.c

```diff
@@ -0,0 +1,88 @@
+/* Copyright (C) 2024 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Mahmoud Maatuq <mahmoudmatook.mm@gmail.com>
+ *
+ */
+
+#include "app-layer.h"
+#include "app-layer-detect-proto.h"
+#include "rust-bindings.h"
+#include "app-layer-imap.h"
+
+static int IMAPRegisterPatternsForProtocolDetection(void)
+{
+    if (AppLayerProtoDetectPMRegisterPatternCI(
+                IPPROTO_TCP, ALPROTO_IMAP, "* OK ", 5, 0, STREAM_TOCLIENT) < 0) {
+        return -1;
+    }
+
+    if (AppLayerProtoDetectPMRegisterPatternCI(
+                IPPROTO_TCP, ALPROTO_IMAP, "* NO ", 5, 0, STREAM_TOCLIENT) < 0) {
+        return -1;
+    }
+
+    if (AppLayerProtoDetectPMRegisterPatternCI(
+                IPPROTO_TCP, ALPROTO_IMAP, "* BAD ", 6, 0, STREAM_TOCLIENT) < 0) {
+        return -1;
+    }
+
+    if (AppLayerProtoDetectPMRegisterPatternCI(
+                IPPROTO_TCP, ALPROTO_IMAP, "* LIST ", 7, 0, STREAM_TOCLIENT) < 0) {
+        return -1;
+    }
+
+    if (AppLayerProtoDetectPMRegisterPatternCI(
+                IPPROTO_TCP, ALPROTO_IMAP, "* ESEARCH ", 10, 0, STREAM_TOCLIENT) < 0) {
+        return -1;
+    }
+
+    if (AppLayerProtoDetectPMRegisterPatternCI(
+                IPPROTO_TCP, ALPROTO_IMAP, "* STATUS ", 9, 0, STREAM_TOCLIENT) < 0) {
+        return -1;
+    }
+
+    if (AppLayerProtoDetectPMRegisterPatternCI(
+                IPPROTO_TCP, ALPROTO_IMAP, "* FLAGS ", 8, 0, STREAM_TOCLIENT) < 0) {
+        return -1;
+    }
+
+    if (AppLayerProtoDetectPMRegisterPatternCI(
+                IPPROTO_TCP, ALPROTO_IMAP, "DONE", 4, 0, STREAM_TOSERVER) < 0) {
```

Comment:
I do not see this `DONE` in the SV PR pcap... but `a0000 CAPABILITY`

Reply:
- [rfc-9051](https://datatracker.ietf.org/doc/html/rfc9051#name-commands-and-responses) DONE with this format is a valid command, 
- I tried to find a pcap that has DONE command but couldn't, If it's super important I might craft one with scapy.

Reply:
But `DONE` does not seem to be the first command meant to be sent, right ?

Reply:
- if what you mean by "first command", is the command that is sent right after tcp handshake so answer is "yes" 
-  regarding this point, I wasn't sure if suricata handles mid stream sessions (no handshake was captured, and/or some other packets in entire session are lost) as if it's the case that suricata ignores such sessions so "DONE" is useless. 
- though, I looked at SMTPRegisterPatternsForProtocolDetection(), and it seems that uses patterns that are not the "first command"  , by the previous definition, .
- not sure, if i can use this as an evidence that suricata handles mid stream sessions, later i'll have a look at tcp part.

Reply:
> what you mean by "first command", is the command that is sent right after tcp handshake 

Yes.

It is the most important part for protocol detection.
We want to handle sessions where we see the beginning.
And if we do not recognize the first bytes as IMAP, we will not try again with the following bytes/commands...
To be complete, as we process the two sides of the connection, we still have a second chance...

Suricata can handle midstream sessions as well. (suricata.yaml option stream.midstream, off by default)

Reply:
- I tested some pcaps with complete sessions and all of them were detected as imap, do you have an example where those patterns fail to detect imap?

Reply:
And this is good.

This is because Suricata detects imap on the `STREAM_TOCLIENT` side.

It is even better to detect both sides.
(You can comment out the line with `DONE` and stil get your SV test passing)

Reply:
- I still didn't completely get what is wrong with those patterns, as they work fine with a complete session and have a good chance to detect mid session. 
- I'll try to go through your concern, and see if what i understood make sense: 
     * the concern is that "DONE" command is not the first client to server command and preferably to have the first client to server command for detection. 
     * in this previous [PR](https://github.com/OISF/suricata/pull/10275) there were comprehensive patterns but processing wise it was costly. 
     * so the tradeoff to have something that is not comprehensive but covers a good range of detection could be acceptable. 
     * the implementation of ftp, smtp patterns reflects that. 
     * removing "DONE" and still works with complete session it's true and i think it might be acceptable to have only one side  such as smtp implementation. 
     * still having "DONE"  is better than removing it as it's a cheap in terms of processing and it increases the chance of mid session detection. 
     * having other client commands won't be as simple as "DONE" as they have "UID" that have dynamic size.

Reply:
> I still didn't completely get what is wrong with those patterns

I do not think there is something wrong with these patterns :-)

I wonder how the `STREAM_TOSERVER` patterns can be improved.
Can there be `capability` case insensitive with some offset and depth ?

> still having "DONE" is better than removing it as it's a cheap in terms of processing and it increases the chance of mid session detection.

But that would require that Suricata sees `DONE` as the first bytes of the flow cf arguments `0` (offset to begin), and `4` depth, so that looks unlikely...
For instance, `SMB` magic happens at every PDU, so we have more chance to see it mid-stream.

Reply:
> > I still didn't completely get what is wrong with those patterns
> 
> I do not think there is something wrong with these patterns :-)
> 
> I wonder how the `STREAM_TOSERVER` patterns can be improved. Can there be `capability` case insensitive with some offset and depth ?
> 
I'm afraid that is not straight forward as the message format is as below 
`
   Example:    C: abcd CAPABILITY
               S: * CAPABILITY IMAP4
               S: abcd OK CAPABILITY completed`
  that **abcd** is **uid** which doesn't have a fixed size. 
  is it possible to use regex in with the current api? 
  e.g `[a-zA-Z0-9]+ CAPABILITY`
> > still having "DONE" is better than removing it as it's a cheap in terms of processing and it increases the chance of mid session detection.
> 
> But that would require that Suricata sees `DONE` as the first bytes of the flow cf arguments `0` (offset to begin), and `4` depth, so that looks unlikely... For instance, `SMB` magic happens at every PDU, so we have more chance to see it mid-stream.

I agree that won't happen every time but it still could happen.

Reply:
> that abcd is uid which doesn't have a fixed size.

Does it have a max size ?

> is it possible to use regex in with the current api?

I do not think we need a regexp : we just need to look for ` capability` in the first n bytes, by setting the right value for depth... To be tested...

Reply:
> > that abcd is uid which doesn't have a fixed size.
> 
> Does it have a max size ?
> 
as per rfc there is no limit for it, thought i saw some unofficial implementation use 10 and others use 20 chars.
> > is it possible to use regex in with the current api?
> 
> I do not think we need a regexp : we just need to look for ` capability` in the first n bytes, by setting the right value for depth... To be tested...
I'm not sure what offset to use at this case

Reply:
offset is minimum uid size, 20 + sizeof(" capability") looks good for depth

Reply:
> offset is minimum uid size, 20 + sizeof(" capability") looks good for depth

Ok I'll submit a new PR with this change.